File: pipeline/src/main/java/dev/mzarnowski/system/pipeline/Pump.java
package dev.mzarnowski.system.pipeline;

import dev.mzarnowski.Disposable;

import java.util.concurrent.Executor;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;

abstract class Pump extends FeedbackComponent implements Disposable {
    private static final int IDLE = 0;
    private static final int RUNNING = 1;
    private static final int QUEUED = 2;
    private static final int COMPLETED = 3;

    protected final Executor executor;
    protected final int batchSize;
    private final Runnable initialTask = this::pump;
    private final AtomicInteger state = new AtomicInteger(IDLE);
    private final AtomicReference<Runnable> task = new AtomicReference<>(initialTask);

    Pump(Pipeline owner) {
        super(owner);
        this.executor = owner.scheduler;
        this.batchSize = owner.batchSize;
        onComplete(() -> state.set(COMPLETED));
    }

    protected abstract void pump();
    protected abstract void drain();

    final void schedule() {
        if (IDLE == state.getAndUpdate(it -> it < QUEUED ? it + 1 : it)) {
            executor.execute(this::iterate);
        }
    }

    final void complete() {
        if (task.compareAndSet(initialTask, this::drain)) {
            schedule();
        }
    }

    @Override
    public final void dispose() {
        if (onComplete != task.getAndSet(onComplete)) {
            schedule();
        }
    }

    private void iterate() {
        task.get().run();

        var is = state.updateAndGet(it -> it == COMPLETED ? COMPLETED : it - 1);
        if (is == RUNNING) executor.execute(this::iterate);
    }
}

File: pipeline/src/main/kotlin/dev/mzarnowski/system/pipeline/Pump.kt
package dev.mzarnowski.system.pipeline

import dev.mzarnowski.Disposable
import java.util.concurrent.atomic.AtomicBoolean
import kotlin.math.min
import java.util.concurrent.ScheduledExecutorService as Scheduler

internal class Pump<A>(private val scheduler: Scheduler, f: (Writer<A>) -> () -> Task.Result) : Disposable,
    Buffer<A>(32) {
    internal val disposed = AtomicBoolean(false)

    internal val writer: Upstream<A> = Upstream(this)
    private val pull: Task = Task(scheduler, this, f(writer)) // TODO make lateinit settable after construction
    private val downstream = mutableMapOf<Reader<A>, Task>()

    internal fun add(reader: Reader<A>, task: Task) {
        downstream[reader] = task
        while (true) {
            val at = (writer.at + writer.available) and mask
            if (reader.at == at) return
            reader.at = at
        }
    }

    fun forEach(f: (A) -> Unit): Disposable {
        val reader = Downstream(this)
        val task = ForEach(scheduler, reader, f)

        add(reader, task)
        return task.also(Task::invoke)
    }

    class Upstream<A>(override val buffer: Pump<A>) : Writer<A>() {
        override fun claim(atLeast: Int, atMost: Int): Int {
            return super.claim(atLeast, atMost).also {
                if (it < atMost) request()
            }
        }

        override fun calculateAvailability(): Int {
            val downstream = buffer.downstream.keys

            if (downstream.isEmpty()) return available
            var available = buffer.size
            downstream.forEach { reader ->
                val distance = buffer.size - (this.at - reader.at)
                if (distance == 0) return 0
                available = min(available, distance)
            }
            return available
        }

        override fun request() = buffer.downstream.values.forEach { it() }
    }

    class Downstream<A>(override val buffer: Pump<A>) : Reader<A>() {
        override fun claim(atLeast: Int, atMost: Int): Int {
            return super.claim(atLeast, atMost).also {
                if (it < atMost) request()
            }
        }

        override fun calculateAvailability(): Int = (buffer.writer.at - this.at)
        override fun request() = buffer.pull()
    }

    override fun dispose() {
        if (disposed.compareAndSet(false, true)) {
            writer.dispose()
            pull.dispose()
            downstream.keys.forEach(Disposable::dispose)
            println("disposed")
        }
    }
}